toc/rfc/rfc-draft-add-cnbs.md

```diff
@@ -0,0 +1,194 @@
+# Meta
+[meta]: #meta
+- Name: Add Cloud Native Buildpacks
+- Start Date: 2023-05-01
+- Author: @gerg
+- Contributing Authors: @ryanmoran, @robdimsdale, @dsabeti, @aramprice, @dsboulder
+- Status: Draft
+- RFC Pull Request: #591
+
+## Summary
+
+[Cloud Native Buildpacks (CNBs)](https://buildpacks.io/), also known as v3
+buildpacks, are the current generation of buildpacks, and offer some
+improvements over the v2 buildpacks that CF Deployment currently uses. The
+Cloud Foundry Foundation already has an implementation of Cloud Native
+Buildpacks via the [Paketo](https://paketo.io/) project, however these CNBs are
+not currently integrated with core CF Deployment.
+
+This RFC proposes adding CNB support to Cloud Foundry and including the Paketo
+buildpacks in CF Deployment.
+
+## Problem
+
+The v2 buildpacks are effectively in maintenance mode and do not receive
+substantial new features. By not integrating with v3 buildpacks, Cloud Foundry
+is missing out on new capabilities like application
+[SBOM](https://en.wikipedia.org/wiki/Software_supply_chain) generation, new
+buildpacks like the Java Native and Web Servers CNBs, as well as any new
+features that are added to the still-actively-developed v3 buildpacks.
+
+## Proposal
+
+### Overview
+
+Cloud Foundry integration with CNBs will be executed over at least two phases.
+
+In the first phase, the Paketo buildpacks will be "shimmed" to work with Cloud
+Foundry's existing v2 buildpack interface. These shimmed Paketo buildpacks will
+be bundled with select [CNB lifecycle](https://github.com/buildpacks/lifecycle)
+binaries and lightweight orchestration binaries. At build time, the Cloud
+Foundry [buildpack lifecycle](https://github.com/cloudfoundry/buildpackapplifecycle)
+will invoke the orchestration binaries, which will map the v2 build process to
+the v3 build process.
+
+In the second phase, unmodified Paketo buildpacks will be natively supported by
+Cloud Foundry. This will be implemented by integrating the buildpack shim logic
+into a Cloud Foundry lifecycle.
+
+Depending on learnings from the first two phases of CNB integration or other
+factors, there may be additional phases of work, however these additional
+phases are not covered by this RFC.
+
+Notably, this proposal does NOT include building or running OCI images on Cloud
+Foundry. This proposal is to use Paketo buildpacks to build and run Cloud
+Foundry droplets. This proposal does not preclude future efforts to integrate
+Cloud Foundry with OCI images, container registries, etc.
+
+### Goals
+
+1. Align CF Deployment with the main focus of the Buildpacks team
+1. Bring new build capabilities to CF Deployment users, including:
+    1. [Application SBOMs](https://paketo.io/docs/howto/sbom/)
+    1. [Web Servers Buildpack](https://github.com/paketo-buildpacks/web-servers)
+    1. [Java Native Buildpack](https://github.com/paketo-buildpacks/java-native-image)
+    1. Easier buildpack customization/composition
```

Comment:
Is there now change required in CF to support this goal (Easier buildpack customization/composition)? In the RFC I couldn't find anything in this direction. Or this is covered by the [multiple-buildpacks support](https://docs.cloudfoundry.org/buildpacks/use-multiple-buildpacks.html)?

Reply:
I don't think there are any necessary changes to CF to get this capability. The main Paketo buildpacks are themselves [composites of multiple other buildpacks](https://github.com/paketo-buildpacks/ruby/blob/main/buildpack.toml), which makes it easier to add/remove sub-buildpacks to meet your custom needs.

That said, this could open the door for more advanced operator controls for buildpacks in the future. For instance, if platform operators could customize buildpacks via platform configuration instead of modifying the buildpacks themselves. Definitely outside the scope of this RFC though.